packages/core/upload/admin/src/components/SelectTree/SelectTree.js

```diff
@@ -0,0 +1,100 @@
+/* eslint-disable react/prop-types */
+
+import React, { useEffect, useState, useMemo } from 'react';
+import styled from 'styled-components';
+import PropTypes from 'prop-types';
+import { components } from 'react-select';
+
+import { Flex } from '@strapi/design-system/Flex';
+import { Icon } from '@strapi/design-system/Icon';
+import { ReactSelect as Select, pxToRem } from '@strapi/helper-plugin';
+import { Typography } from '@strapi/design-system/Typography';
+import ChevronUp from '@strapi/icons/ChevronUp';
+import ChevronDown from '@strapi/icons/ChevronDown';
+
+import flattenTree from './utils/flattenTree';
+
+const ToggleButton = styled.button`
+  align-self: flex-end;
+  margin-left: auto;
+`;
+
+const hasParent = option => !option.parent;
+
+const hasParentOrMatchesValue = (option, value) =>
+  option.value === value || option.parent === value;
+
+const SelectTree = ({ options: defaultOptions, maxDisplayDepth, ...props }) => {
+  const flatDefaultOptions = useMemo(() => flattenTree(defaultOptions), [defaultOptions]);
+  const toplevelDefaultOptions = useMemo(() => flatDefaultOptions.filter(hasParent), [
+    flatDefaultOptions,
+  ]);
+  const [options, setOptions] = useState(toplevelDefaultOptions);
+  const [openValues, setOpenValues] = useState([]);
+
+  useEffect(() => {
+    if (openValues.length === 0) {
+      setOptions(toplevelDefaultOptions);
+    }
+
+    openValues.forEach(value => {
+      const filtered = flatDefaultOptions.filter(
+        option => hasParentOrMatchesValue(option, value) || hasParent(option)
+      );
+
+      setOptions(filtered);
+    });
+  }, [openValues, flatDefaultOptions, toplevelDefaultOptions]);
+
+  function handleToggle(e, value) {
+    e.preventDefault();
+    e.stopPropagation();
+
+    if (openValues.includes(value)) {
+      setOpenValues(prev => prev.filter(prevData => prevData !== value));
+    } else {
+      setOpenValues(prev => [...prev, value]);
+    }
+  }
+
+  const CustomOption = ({ children, data, ...props }) => {
```

Comment:
You should create a component for this that is out of the scope of the Parent one.